### ISSUE TITLE
Can't compile for C99 due to strdup and uint

### ISSUE THREAD

**Issue by totszwai** [CESNET/libyang#483]
Trying to use libyang for embedded system, when compiling against arm-linux toolchain, the macro **LY_ARRAY_MAX** in **libyang/src/tree_schema.h:171** asks for **ULLONG_MAX**. And ULLONG_MAX in limits.h is only available when is C99 or higher.

usr/lib/gcc/arm-buildroot-linux-uclibcgnueabi/4.7.2/include-fixed/limits.h
```
#if defined (__STDC_VERSION__) && __STDC_VERSION__ >= 199901L
/* Minimum and maximum values a `signed long long int' can hold.  */
# undef LLONG_MIN
# define LLONG_MIN (-LLONG_MAX - 1LL)
# undef LLONG_MAX
# define LLONG_MAX __LONG_LONG_MAX__

/* Maximum value an `unsigned long long int' can hold.  (Minimum is 0).  */
# undef ULLONG_MAX
# define ULLONG_MAX (LLONG_MAX * 2ULL + 1ULL)
#endif
```

usr/arm-buildroot-linux-uclibcgnueabi/sysroot/usr/include/limits.h
```
#  ifdef __USE_ISOC99

/* Minimum and maximum values a `signed long long int' can hold.  */
#   define LLONG_MAX    9223372036854775807LL
#   define LLONG_MIN    (-LLONG_MAX - 1LL)

/* Maximum value an `unsigned long long int' can hold.  (Minimum is 0.)  */
#   define ULLONG_MAX   18446744073709551615ULL

#  endif /* ISO C99 */
```

However if I set it to C99 with `-std=c99`, it will complaint because of `strdup` being used in several places in libyang. And I get errors in several places:

> libyang-0.14-r1/src/parser_yang.h:173:88: error: unknown type name 'uint'
> libyang-0.14-r1/src/parser_yang_bis.c:3633:5: warning: implicit declaration of function 'strdup' [-Wimplicit-function-declaration]

What is the proper way to compile libyang for arm-linux?

**Comment by rkrejci** [comment]
Hi, try `-std=gnu99`. In devel, we already use `-std=gnu11` and require gcc >= 4.9 because of use of `_Generic` statement.

**Comment by totszwai** [comment]
Is there a way to disable extension? 

> extensions.c:128:44: error: 'RTLD_NOLOAD' undeclared

We use glibc 2.12, don't think we could change our build system's configuration anytime soon.

**Comment by rkrejci** [comment]
No, but maybe it will be solved in CESNET/Netopeer2#138

**Comment by totszwai** [comment]
For now I patched out the block of code that was in CESNET/Netopeer2#138 and worked for glibc 2.12. I guess that check is nice to have.

**Comment by mislavn** [comment]
@totszwai the PR https://github.com/CESNET/libyang/pull/484 should fix your issue.

**Comment by rkrejci** [comment]
Hi, glibc 2.12 is 8 years old, do you have any good reason to use it?

**Comment by totszwai** [comment]
If i can change the build system, I would, but I have no control over it and apparently there are dozens of applications (old) that gets build by it.

**Comment by rkrejci** [comment]
I'm afraid that I'm not going to clutter the code because of support for the not maintained glibc version. I'm going to fix the issue described in CESNET/Netopeer2#138, but missing `RTLD_NOLOAD` is different thing.

Note that the proposed walkaround mentioned in CESNET/Netopeer2#138 is not perfect - it introduces possible memory leaks in some scenarios of creating and destroying libyang contexts. You would need to change more code to have a correct solution for glibc 2.12.

**Comment by rkrejci** [comment]
well, at the end I was able to find a generic solution which avoids `RTLD_NOLOAD`. Please test the proposed patch.
